### PR TITLE
Scrapper, Maxim Matveev - 21FPL1

### PR DESCRIPTION
**Pull request by yourMasya**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Welcome to the CTLR course! 
Reviews will appear every Wednesday. 
Make sure to publish relevant changes by that day of the week. 
Good luck!

**Review by marina-kaz — COMMENTED** [review]
Nothing to review as of today

**Review by marina-kaz — COMMENTED** [review]
Nothing substantial to review at this point

**Review by marina-kaz — COMMENTED** [review]
Nothing to review today

**Review by marina-kaz — COMMENTED** [review]
Nothing to review today

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
No progress here

**Comment by yourMasya** [comment]
> No progress here
Yes.ть

**Comment by yourMasya** [comment]
> No progress here

Yes.ть

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -13,70 +72,121 @@ def __init__(self, path_to_config: Path) -> None:
...
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        config_dto = self._extract_config_content()
+
+        if not isinstance(config_dto.seed_urls, list):
+            raise IncorrectSeedURLError
+
+        for url in config_dto.seed_urls:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:108]: what if seed_urls is empty?

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -13,70 +72,121 @@ def __init__(self, path_to_config: Path) -> None:
...
 
     def get_headless_mode(self) -> bool:
         """
         Retrieve whether to use headless mode
         """
-        pass
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:179]: consider also specifying the encoding of a response in this function

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -90,25 +200,47 @@ def __init__(self, config: Config) -> None:
...
+    # seed URLs and stores newly collected ones,
+    # while all the extraction is performed via protected
+    # _extract_url method.
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        url = article_bs.get('href')
+        # url = re.compile(r'(?<=href=")[\w./:]+(?=")')
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: do not forget to remove unused code before the hand-in

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -90,25 +200,47 @@ def __init__(self, config: Config) -> None:
...
         """
         Finds articles
         """
-        pass
+        for seed_url in self._seed_urls:
+            res = make_request(seed_url, self._config)
+            html = BeautifulSoup(res.content, 'lxml')
+            # print(html.find_all('a', class_="news-list-card"))
+            for paragraph in html.find_all('a', {'class': "news-list-card"}):
+                if len(self.urls) >= self._config.get_num_articles():
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py:229]: consider moving this check 
as of now, it looks like you'll have to make one extra request which takes additional time etc.

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -120,45 +252,109 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 
     def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         """
         Finds text of article
         """
-        pass
+        article_content = article_soup.find("div", {'class': "article-body__content"})
+        paragraphs = article_content.find_all("p")
+        self.article.text = ' '.join(paragraph.text.strip() for paragraph in paragraphs)
+        print(self.article.text)
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: is this for self check? 
consider removing this print, as in case of 100 texts your console output will end up very busy...

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -120,45 +252,109 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
-        pass
+        title = article_soup.find('h1', class_="article-body__title")
+        if title:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: this condition looks like there can be cases when the title field will remain unfilled 
it is unacceptable, title field must always be filled

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -120,45 +252,109 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            self.article.title = title.text
+        date = article_soup.find('div', class_="single-header__time")
+        if date:
```
Comment (marina-kaz) [inline lab_5_scrapper/scrapper.py]: same for the date

File: lab_6_pipeline/pipeline.py [hunk 8]
```diff
@@ -1,187 +0,0 @@
-"""
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:1]: why did you need to delete the stubs for the next lab? 
if there is no necessity, please bring it back

**Review by marina-kaz — COMMENTED** [review]
Almost done!
Some minor fixes left 

Please remove unused code and unnecessary printing

**Review by marina-kaz — COMMENTED** [review]
No changes since my last review

**Review by marina-kaz — COMMENTED** [review]
No changes since my last review

**Review by marina-kaz — COMMENTED** [review]
No changes since last review

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (yourMasya) [reply lab_5_scrapper/scrapper.py:108]: resolved

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (yourMasya) [reply lab_5_scrapper/scrapper.py:179]: I don't really understand what is required. There is another method retrieving encoding to use during parsing. Or should I implement standard isinstance(self._headless_mode, bool) check and return None if the condition is unsatisfied?

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (yourMasya) [reply lab_5_scrapper/scrapper.py]: removed comments and prints

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (yourMasya) [reply lab_5_scrapper/scrapper.py]: deleted

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (yourMasya) [reply lab_5_scrapper/scrapper.py:179]: Returning None is non-relevant here because of bool type in function's requirments

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (marina-kaz) [reply lab_5_scrapper/scrapper.py]: to be pushed

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (marina-kaz) [reply lab_5_scrapper/scrapper.py]: to be pushed
